### PR TITLE
Striped layouts in cephfs-data-scan

### PR DESCRIPTION
**Pull request by jcsp**

### CODE REVIEW

**Comment by jcsp** [comment]
The test for this is https://github.com/ceph/ceph-qa-suite/pull/526

**Comment by ukernel** [comment]
there are compilation errors, please update

**Comment by jcsp** [comment]
Oops, this lost a commit during rebase.  Fixed.

File: src/tools/cephfs/DataScan.cc [hunk 1]
```diff
@@ -515,8 +525,80 @@ int DataScan::scan_inodes()
...
+      } else {
+        // Striped file: need to examine the last fl_stripe_count objects
+        // in the file to determine the size.
+
+        // How many complete (i.e. not last stripe) objects?
+        uint64_t complete_objs = 0;
+        if (accum_res.ceiling_obj_index > guessed_layout.fl_stripe_count - 1) {
+          // e.g. we've got 8 objects, ceiling index is 7, stripe_count is 4,
+          // complete object should then be 4 (7 + 1 - 4)
+          complete_objs = accum_res.ceiling_obj_index + 1 - guessed_layout.fl_stripe_count;
```
Comment (ukernel) [inline src/tools/cephfs/DataScan.cc]: This seems incorrect. If there are 9 objects, ceiling index is 8, stripe_count is 4, complete object should be 8. I think complete_objs should be ceiling_obj_index / fl_stripe_count * fl_stripe_count.

File: src/tools/cephfs/DataScan.cc [hunk 2]
```diff
@@ -515,8 +525,80 @@ int DataScan::scan_inodes()
...
+        }
+        if (r != 0 && r != -ENOENT) {
+          derr << "Unexpected error checking size of ino 0x" << std::hex
+               << obj_name_ino << std::dec << ": " << cpp_strerror(r) << dendl;
+          continue;
+        }
+
+        file_size = complete_objs * guessed_layout.fl_object_size
+                   + incomplete_size;
+      }
```
Comment (ukernel) [inline src/tools/cephfs/DataScan.cc]: incomplete_size also seems incorrect. 

max_offset_of_an_object = (osize - 1) / fl_stripe_unit * (fl_stripe_unit * fl_stripe_count) + (i % fl_stripe_count) * fl_stripe_unit + (osize - 1) % fl_stripe_unit + 1

incomplete_size = max(incomplete_size, max_offset_of_an_object)

**Comment by jcsp** [comment]
Thanks for correcting the maths: I've updated this PR, and added more tests to the ceph-qa-suite one.

File: src/tools/cephfs/DataScan.cc [hunk 3]
```diff
@@ -515,8 +525,92 @@ int DataScan::scan_inodes()
...
+        // and accumulate the maximum
+        for (uint64_t i = complete_objs; i < complete_objs + partial_objs; ++i) {
+          char buf[60];
+          snprintf(buf, sizeof(buf), "%llx.%08llx",
+              (long long unsigned)obj_name_ino, (long long unsigned)i);
+
+          uint64_t osize(0);
+          time_t omtime(0);
+          r = data_io.stat(std::string(buf), &osize, &omtime);
+          if (r == 0) {
```
Comment (ukernel) [inline src/tools/cephfs/DataScan.cc]: need check if osize > 0

**Pull request merged**